### PR TITLE
[Spec] DSpark support PD and dp-attention

### PR DESCRIPTION
**Pull request by zhangxiaolei123456**
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

Our DSpark implementation is based on the baseline from upstream SGLang PR https://github.com/sgl-project/sglang/pull/29538 On top of that baseline, we extended DSpark to support DP attention correctly.

DSpark with DP attention keeps all DP ranks participating in the decode, draft, refine, and verify collectives, even when a local rank has an empty batch. Empty ranks are handled with dummy padded metadata and then trimmed back to their real local batch size after collective-sensitive stages.

For correctness, DSpark refinement must compute logits against the local DP request only. Therefore, when DP attention is enabled, DSpark also enables DP LM head, and the refinement path only gathers vocab shards across the real TP group, not across different DP requests.

<!-- Describe the purpose and goals of this pull request. -->

## Modifications

<!-- Detail the changes made in this pull request. -->

## Accuracy Tests
Command
```
SGLANG_DSPARK_FP8_WQKV_STACK=1 NCCL_SOCKET_IFNAME=eth0 NCCL_IB_DISABLE=0 SGLANG_DSV4_FP4_EXPERTS=1 GLOO_SOCKET_IFNAME=eth0 python -m sglang.launch_server --model-path /data00/models/DeepSeek-V4-Flash-DSpark --speculative-dspark-confidence-threshold 0.0 --moe-runner-backend flashinfer_mxfp4 --mem-fraction-static 0.85 --speculative-algo DSPARK --speculative-num-steps 1 --speculative-eagle-topk 1 --speculative-num-draft-tokens 5 --tp 8 --cuda-graph-max-bs 256 --max-running-requests 256 --enable-metrics --host 0.0.0.0 --port 30300 --dp-size 8 --enable-dp-attention

NCCL_SOCKET_IFNAME=eth0 NCCL_IB_DISABLE=0 SGLANG_DSV4_FP4_EXPERTS=1 GLOO_SOCKET_IFNAME=eth0 python -m sglang.launch_server --model-path /data00/models/DeepSeek-V4-Flash --moe-runner-backend flashinfer_mxfp4 --mem-fraction-static 0.85 --speculative-algo EAGLE --speculative-num-steps 4 --speculative-eagle-topk 1 --speculative-num-draft-tokens 5 --tp 8 --cuda-graph-max-bs 32 --max-running-requests 256 --enable-metrics --host 0.0.0.0 --port 30300 --dp-size 8 --enable-dp-attention
```

Mode | GSM8K | MMLU | GPQA 
-- | -- | -- | -- 
DSpark+DP-attention+cuda graph | 0.951 | 0.886 | 0.898


<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Speed Tests and Profiling
PR | Hardward | input | output | chunked-prefill-size | QPS | max_concurrency | num promot | TPOT(ms) | throughput(tokens/s/gpu)
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
DSpark | H20 | 3500 | 1500 | 8192 | 40 | 128 | 512 | 23.71 | 602
DSpark+DP-attention+cuda graph | H20 | 3500 | 1500 | 8192 | 40 | 128 | 512 | 17.6 | 814
EAGLE | H20 | 3500 | 1500 | 8192 | 40 | 128 | 512 | 24.90 | 571


<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.





















































































































































































































































































































































































































































































































































































<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:x: **Missing `run-ci` label** -- add it to run CI tests.<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:x: **Blocked** -- `run-ci` is required first.<!-- slot:pr-test-extra:end -->
<!-- pr-states:end -->

### CODE REVIEW

**Comment by zhangxiaolei123456** [comment]
@BJWang-ant, MTP=1 is --speculative-num-steps=1?  For DSpark, `--speculative-num-steps 1` is more of a placeholder or compatibility parameter. Adjust the DSpark length; you should adjust --speculative-num-draft-tokens

'--speculative-num-steps 1 --speculative-eagle-topk 1 --speculative-num-draft-tokens 5'

**Comment by zhangxiaolei123456** [comment]
@BJWang-ant EAGLE is --speculative-num-steps 4 --speculative-eagle-topk 1 --speculative-num-draft-tokens 5

**Comment by zhangxiaolei123456** [comment]
YES

**Comment by lyon0321** [comment]
May I ask, is there a performance comparison for Eagle + DP attention?

**Comment by zhangxiaolei123456** [comment]
My data is just Eagle + DP attention @lyon0321

**Comment by hnyls2002** [comment]
Thanks @zhangxiaolei123456! This builds on the #29538 base which never merged - native DSpark landed via #30261 with a different architecture (dspark_components/), including DP-attention tier handling and PD-path integration, so the files this PR extends no longer exist on main. Closing along with #30513; broader parallelism coverage is tracked in #30344 and contributions on top of the merged stack are very welcome.

**Comment by yiakwy-xpu-ml-framework-team** [comment]
Hi @zhangxiaolei123456 

When DP is enabled, there is a regression in test score :

Mode | GSM8K
-- | --
No Spec | 0.970
DSpark (Default T=1.0) | 0.990
DSpark+DP-attention+cuda graph | 0.951

For 8K questions in GSM8K, upto 160 to 320 questions answered differently.

Could we sample 1 or more questions to see how they were ansered differently ?
